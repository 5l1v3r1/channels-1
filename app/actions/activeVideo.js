// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
