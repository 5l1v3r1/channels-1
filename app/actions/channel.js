// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function selectChannel(id: string): Action {
  return {
    type: actionTypes.SELECT_CHANNEL,
    payload: id
  };
}
