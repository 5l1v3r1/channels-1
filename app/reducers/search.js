// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';
import VideoType from '../dataTypes/videoType';
import { createIsFetching, createError } from './common';

function query(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.UPDATE_SEARCH_QUERY:
      return payload;
    default:
      return state;
  }
}

function results(state: VideoType[] = [], { type, payload }: Action): VideoType[] {
  switch (type) {
    case actionTypes.FETCH_SEARCH_SUCCESS:
      return [
        ...state,
        ...payload.items.map(item => VideoType.from(item))
      ];
    case actionTypes.CLEAR_SEARCH_RESULTS:
      return [];
    default:
      return state;
  }
}

function pageToken(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_SEARCH_SUCCESS:
      return payload.nextPageToken;
    case actionTypes.CLEAR_SEARCH_RESULTS:
      return '';
    default:
      return state;
  }
}

const search = combineReducers({
  query,
  results,
  pageToken,
  isFetching: createIsFetching('SEARCH'),
  error: createError('SEARCH')
});

export default search;
