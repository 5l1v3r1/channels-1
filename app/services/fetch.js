// @flow
export const Methods = {
  GET: 'GET',
  POST: 'POST'
};

export function authRequest(uri: string): Promise<any> {
  const [url, body] = uri.split('?');
  const init = {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body
  };
  return fetch(url, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function getRequest(uri: string): Promise<any> {
  const init = {
    method: Methods.GET,
    headers: { 'Content-Type': 'application/json' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function postRequest(uri: string, body?: {}): Promise<any> {
  const init = body ?
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  } :
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res;
    });
}
