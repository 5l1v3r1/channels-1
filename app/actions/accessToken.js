// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
