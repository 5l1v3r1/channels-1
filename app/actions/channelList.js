// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
