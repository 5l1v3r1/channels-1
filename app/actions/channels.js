// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { token } = getState().accessToken;
    const subscriptions = fetchSubscriptions(token);
    return subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(token, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          return dispatch(receiveChannels(json.items));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

export function fetchSubscriptions(token: string): Promise<any> {
  const uri = getSubscriptionsUri(token);
  return apiRequest(uri, Methods.GET);
}

export function requestChannels(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_REQUEST
  };
}

export function receiveChannels(items: any): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_SUCCESS,
    payload: items
  };
}

export function channelsError(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_FAILURE
  };
}

export function updateActiveChannel(channelId: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { activeId } = getState().channels;
    if (activeId !== channelId) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_CHANNEL,
        payload: channelId
      });
    }
  };
}
