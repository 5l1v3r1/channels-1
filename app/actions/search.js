// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/types';
import type { VideoItem } from '../dataTypes/videoType';
import { getRequest } from '../services/fetch';
import { getVideosUri, getVideoIdsUri } from '../services/uriGenerator';

export function fetchSearch(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(fetchSearchRequest());
    const params = {
      channelId: getState().channels.activeId,
      q: getState().search.query,
      pageToken: getState().search.pageToken,
      order: 'relevance'
    };
    return fetchVideoIds(params)
      .then(
        res => {
          const { videoIds, nextPageToken } = res;
          const uri = getVideosUri(videoIds);
          return getRequest(uri)
            .then(
              json => dispatch(fetchSearchSuccess(json.items, nextPageToken)),
              () => dispatch(fetchSearchFailure())
            );
        },
        () => dispatch(fetchSearchFailure())
      );
  };
}

export function fetchVideoIds(params: any): Promise<any> {
  const uri = getVideoIdsUri(params);
  return getRequest(uri)
    .then(
      json => Promise.resolve({
        videoIds: json.items.map(item => item.id.videoId).join(),
        nextPageToken: json.nextPageToken || params.pageToken
      }),
      () => Promise.reject()
    );
}

export function fetchSearchRequest(): Action {
  return {
    type: actionTypes.FETCH_SEARCH_REQUEST
  };
}

export function fetchSearchSuccess(items: VideoItem[], nextPageToken: string): Action {
  return {
    type: actionTypes.FETCH_SEARCH_SUCCESS,
    payload: {
      items,
      nextPageToken
    }
  };
}

export function fetchSearchFailure(): Action {
  return {
    type: actionTypes.FETCH_SEARCH_FAILURE
  };
}

export function updateSearch(query: string): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(clearSearchResults());
    dispatch(updateSearchQuery(query));
  };
}

export function clearSearch(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(updateSearchQuery(''));
    dispatch(clearSearchResults());
  };
}

export function updateSearchQuery(query: string): Action {
  return {
    type: actionTypes.UPDATE_SEARCH_QUERY,
    payload: query
  };
}

export function clearSearchResults(): Action {
  return {
    type: actionTypes.CLEAR_SEARCH_RESULTS
  };
}
