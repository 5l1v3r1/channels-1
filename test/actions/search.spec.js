/* eslint promise/always-return: 0 */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as searchActions from '../../app/actions/search';
import * as fetch from '../../app/services/fetch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../../app/services/uriGenerator');

describe('search actions', () => {

  let store;
  beforeEach(() => {
    store = mockStore({
      accessToken: {
        token: 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg'
      },
      channels: {
        activeId: 'UCyIe-61Y8C4_o-zZCtO4ETQ'
      },
      search: {
        pageToken: ''
      }
    });
  });

  it('creates FETCH_SEARCH_SUCCESS when fetching video ids has resolved with undefined', () => {
    fetch.apiRequest = jest.fn(() => Promise.resolve({}));

    return store.dispatch(searchActions.fetchSearch()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_SEARCH_SUCCESS when fetching video ids has resolved with null', () => {
    fetch.apiRequest = jest.fn(() => Promise.resolve({ items: null }));

    return store.dispatch(searchActions.fetchSearch()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_SEARCH_FAILURE when fetching video ids has been rejected', () => {
    fetch.apiRequest = jest.fn(() => Promise.reject());

    return store.dispatch(searchActions.fetchSearch()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_SEARCH_SUCCESS when fetching videos has been resolved', () => {
    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            id: {
              videoId: 'M8l2aGMjKHI'
            }
          },
          {
            id: {
              videoId: 'GcSACxUbqtg'
            }
          }
        ],
        nextPageToken: 'CBQQAA'
      }))
      .mockReturnValueOnce(Promise.resolve({ items: ['M8l2aGMjKHI', 'GcSACxUbqtg'] }));

    return store.dispatch(searchActions.fetchSearch()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('creates FETCH_SEARCH_FAILURE when fetching videos has been rejected', () => {
    fetch.apiRequest = jest.fn()
      .mockReturnValueOnce(Promise.resolve({
        items: [
          {
            id: {
              videoId: 'M8l2aGMjKHI'
            }
          },
          {
            id: {
              videoId: 'GcSACxUbqtg'
            }
          }
        ],
        nextPageToken: 'CBQQAA'
      }))
      .mockReturnValueOnce(Promise.reject());

    return store.dispatch(searchActions.fetchSearch()).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  it('should create an action to request search', () => {
    expect(searchActions.requestSearch()).toMatchSnapshot();
  });

  it('should create an action to receive search', () => {
    const items = ['M8l2aGMjKHI', 'GcSACxUbqtg'];
    const nextPageToken = 'CBQQAA';
    expect(searchActions.receiveSearch(items, nextPageToken)).toMatchSnapshot();
  });

  it('should create an action to handle an error', () => {
    expect(searchActions.searchError()).toMatchSnapshot();
  });

  it('should create actions to update search', () => {
    store.dispatch(searchActions.updateSearch('a'));
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should create actions to clear search', () => {
    store.dispatch(searchActions.clearSearch());
    expect(store.getActions()).toMatchSnapshot();
  });
});
