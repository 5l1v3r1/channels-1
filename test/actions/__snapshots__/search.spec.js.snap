// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search actions creates FETCH_SEARCH_FAILURE when fetching video ids has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_SEARCH_REQUEST",
  },
  Object {
    "type": "FETCH_SEARCH_FAILURE",
  },
]
`;

exports[`search actions creates FETCH_SEARCH_FAILURE when fetching videos has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_SEARCH_REQUEST",
  },
  Object {
    "type": "FETCH_SEARCH_FAILURE",
  },
]
`;

exports[`search actions creates FETCH_SEARCH_SUCCESS when fetching video ids has resolved with null 1`] = `
Array [
  Object {
    "type": "FETCH_SEARCH_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [],
      "nextPageToken": "",
    },
    "type": "FETCH_SEARCH_SUCCESS",
  },
]
`;

exports[`search actions creates FETCH_SEARCH_SUCCESS when fetching video ids has resolved with undefined 1`] = `
Array [
  Object {
    "type": "FETCH_SEARCH_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [],
      "nextPageToken": "",
    },
    "type": "FETCH_SEARCH_SUCCESS",
  },
]
`;

exports[`search actions creates FETCH_SEARCH_SUCCESS when fetching videos has been resolved 1`] = `
Array [
  Object {
    "type": "FETCH_SEARCH_REQUEST",
  },
  Object {
    "payload": Object {
      "items": Array [
        "M8l2aGMjKHI",
        "GcSACxUbqtg",
      ],
      "nextPageToken": "CBQQAA",
    },
    "type": "FETCH_SEARCH_SUCCESS",
  },
]
`;

exports[`search actions should create actions to clear search 1`] = `
Array [
  Object {
    "payload": "",
    "type": "UPDATE_SEARCH_QUERY",
  },
  Object {
    "type": "CLEAR_SEARCH_RESULTS",
  },
]
`;

exports[`search actions should create actions to update search 1`] = `
Array [
  Object {
    "type": "CLEAR_SEARCH_RESULTS",
  },
  Object {
    "payload": "a",
    "type": "UPDATE_SEARCH_QUERY",
  },
]
`;

exports[`search actions should create an action to handle an error 1`] = `
Object {
  "type": "FETCH_SEARCH_FAILURE",
}
`;

exports[`search actions should create an action to receive search 1`] = `
Object {
  "payload": Object {
    "items": Array [
      "M8l2aGMjKHI",
      "GcSACxUbqtg",
    ],
    "nextPageToken": "CBQQAA",
  },
  "type": "FETCH_SEARCH_SUCCESS",
}
`;

exports[`search actions should create an action to request search 1`] = `
Object {
  "type": "FETCH_SEARCH_REQUEST",
}
`;
